src/main/scala/play-json.scala

```diff
@@ -444,6 +444,82 @@ This can be caused by https://issues.scala-lang.org/browse/SI-7046 which can onl
     t
   }
 
+  def formatHinted[T: c.WeakTypeTag](enclosed: Tree): Tree = {
+    assertClass[T]()
+    val T = c.weakTypeOf[T]
+
+    val t =
+      q"""
+      {
+        import $pjson._
+        import $pkg._
+        new Format[$T] {
+          private val _delegateFormat: Format[$T] = $enclosed
+          private val _hints: Hints = implicitly[Hints]
+
+          def reads(json: JsValue): JsResult[$T] = {
+            (json \ _hints.field).asOpt[String] match {
+              case Some(hint) if _hints.isHintForClass(hint, classOf[$T]) => _delegateFormat.reads(json)
+              case Some(unexpectedHint) => JsError(s"Cannot deserialize type [$${classOf[$T]}] from json with type-hint [$${unexpectedHint}]")
+              case None => JsError(s"Cannot deserialize type [$${classOf[$T]}] from json without a type-hint field [_hints.field]")
+            }
+          }
+
+          def writes(value: $T): JsValue = _delegateFormat.writes(value) match {
+            case obj: JsObject => obj + (_hints.field, JsString(_hints.hintFor(classOf[$T])))
+            case nonObj => throw new Exception(s"Cannot put type-hint to $${nonObj.getClass.getSimpleName} produced by Format["+${Literal(Constant(T.toString))}+"]. Hinting supported only for Formats that write JsObjects")
+          }
+        }
+      }
+      """
+//    debugMacro(t)
+    t
+  }
+
+  def formatHintedSealed[T: c.WeakTypeTag]: Tree = {
+    assertSealedAbstract[T]
+
+    val T = c.weakTypeOf[T]
+    val subs = T.typeSymbol.asClass.knownDirectSubclasses.toVector
+    if (subs.isEmpty)
+      c.error(c.enclosingPosition,s"""
+No child classes found for $T. If there clearly are child classes,
+try moving the call into a separate file, a sibbling package, a separate sbt sub project or else.
+This can be caused by https://issues.scala-lang.org/browse/SI-7046 which can only be avoided by manually moving the call.""")
```

Comment:
Please take me off this list. I don't know you
On Mar 11, 2016 8:33 AM, "Dmytro Aleksandrov" <notifications@github.com>
wrote:

> In src/main/scala/play-json.scala
> <https://github.com/cvogt/play-json-extensions/pull/22#discussion_r55844073>
> :
>
> > +      }
> > +      """
> > +//    debugMacro(t)
> > +    t
> > +  }
> > +
> > +  def formatHintedSealed[T: c.WeakTypeTag]: Tree = {
> > +    assertSealedAbstract[T]
> > +
> > +    val T = c.weakTypeOf[T]
> > +    val subs = T.typeSymbol.asClass.knownDirectSubclasses.toVector
> > +    if (subs.isEmpty)
> > +      c.error(c.enclosingPosition,s"""
> > +No child classes found for $T. If there clearly are child classes,
> > +try moving the call into a separate file, a sibbling package, a separate sbt sub project or else.
> > +This can be caused by https://issues.scala-lang.org/browse/SI-7046 which can only be avoided by manually moving the call.""")
>
> ok, will fix that.
> BTW, do you have preferred way of mitigating knownDirectSubclasses issue?
> I've noticed several times that moving formatSealed call to sibling
> package is not enough. While I didn't want to make separate module just for
> that, I've come to solution with additional sbt configuration (like
> early-compile) where all model classes live and json formats are compiled
> in normal sbt compile
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cvogt/play-json-extensions/pull/22/files#r55844073>.
>